compiler/rustc_typeck/src/check/method/probe.rs

```diff
@@ -1102,8 +1103,75 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
 
     fn pick_core(&mut self) -> Option<PickResult<'tcx>> {
         let steps = self.steps.clone();
+        let mut unstable_candidates = Vec::new();
 
         // find the first step that works
+        let pick = steps
+            .iter()
+            .filter(|step| {
+                debug!("pick_core: step={:?}", step);
+                // skip types that are from a type error or that would require dereferencing
+                // a raw pointer
+                !step.self_ty.references_error() && !step.from_unsafe_deref
+            })
+            .flat_map(|step| {
+                let InferOk { value: self_ty, obligations: _ } = self
+                    .fcx
+                    .probe_instantiate_query_response(
+                        self.span,
+                        &self.orig_steps_var_values,
+                        &step.self_ty,
+                    )
+                    .unwrap_or_else(|_| {
+                        span_bug!(self.span, "{:?} was applicable but now isn't?", step.self_ty)
+                    });
+                self.pick_by_value_method(step, self_ty, Some(&mut unstable_candidates)).or_else(
+                    || {
+                        self.pick_autorefd_method(
+                            step,
+                            self_ty,
+                            hir::Mutability::Not,
+                            Some(&mut unstable_candidates),
+                        )
+                        .or_else(|| {
+                            self.pick_autorefd_method(
+                                step,
+                                self_ty,
+                                hir::Mutability::Mut,
+                                Some(&mut unstable_candidates),
+                            )
+                        })
+                        .or_else(|| {
+                            self.pick_const_ptr_method(
+                                step,
+                                self_ty,
+                                Some(&mut unstable_candidates),
+                            )
+                        })
```

Comment:
Would be ideal if we could deduplicate this section of code between code picking a stable method and one picking an unstable one. Otherwise it seems like we'd be prone to changing one but not the other and introducing an inconsistency in the future.